File: ios/Sources/RynBridgeMedia/MediaModule.swift
import Foundation
import RynBridge

public struct MediaModule: BridgeModule, Sendable {
    public let name = "media"
    public let version = "0.1.0"
    public let actions: [String: ActionHandler]

    public init(provider: MediaProvider) {
        actions = [
            "playAudio": { payload in
                guard let source = payload["source"]?.stringValue else {
                    throw RynBridgeError(code: .invalidMessage, message: "Missing required field: source")
                }
                let loop = payload["loop"]?.boolValue ?? false
                let volume = payload["volume"]?.doubleValue ?? 1.0
                let playerId = try await provider.playAudio(source: source, loop: loop, volume: volume)
                return ["playerId": .string(playerId)]
            },
            "pauseAudio": { payload in
                guard let playerId = payload["playerId"]?.stringValue else {
                    throw RynBridgeError(code: .invalidMessage, message: "Missing required field: playerId")
                }
                try await provider.pauseAudio(playerId: playerId)
                return [:]
            },
            "stopAudio": { payload in
                guard let playerId = payload["playerId"]?.stringValue else {
                    throw RynBridgeError(code: .invalidMessage, message: "Missing required field: playerId")
                }
                try await provider.stopAudio(playerId: playerId)
                return [:]
            },
            "getAudioStatus": { payload in
                guard let playerId = payload["playerId"]?.stringValue else {
                    throw RynBridgeError(code: .invalidMessage, message: "Missing required field: playerId")
                }
                let status = try await provider.getAudioStatus(playerId: playerId)
                return status.toPayload()
            },
            "startRecording": { payload in
                let format = payload["format"]?.stringValue ?? "m4a"
                let quality = payload["quality"]?.stringValue ?? "medium"
                let recordingId = try await provider.startRecording(format: format, quality: quality)
                return ["recordingId": .string(recordingId)]
            },
            "stopRecording": { payload in
                guard let recordingId = payload["recordingId"]?.stringValue else {
                    throw RynBridgeError(code: .invalidMessage, message: "Missing required field: recordingId")
                }
                let result = try await provider.stopRecording(recordingId: recordingId)
                return result.toPayload()
            },
            "pickMedia": { payload in
                let type = payload["type"]?.stringValue ?? "any"
                let multiple = payload["multiple"]?.boolValue ?? false
                let files = try await provider.pickMedia(type: type, multiple: multiple)
                return ["files": .array(files.map { .object($0.toPayload()) })]
            },
        ]
    }
}

File: android/media/src/main/kotlin/io/rynbridge/media/MediaModule.kt
package io.rynbridge.media

import android.content.Context
import io.rynbridge.core.*

class MediaModule(provider: MediaProvider) : BridgeModule {
    constructor(context: Context) : this(DefaultMediaProvider(context))

    override val name = "media"
    override val version = "0.1.0"
    override val actions: Map<String, ActionHandler> = mapOf(
        "playAudio" to { payload ->
            val source = payload["source"]?.stringValue
                ?: throw RynBridgeError(code = ErrorCode.INVALID_MESSAGE, message = "Missing required field: source")
            val loop = payload["loop"]?.boolValue ?: false
            val volume = payload["volume"]?.doubleValue ?: 1.0
            val playerId = provider.playAudio(source, loop, volume)
            mapOf("playerId" to BridgeValue.string(playerId))
        },
        "pauseAudio" to { payload ->
            val playerId = payload["playerId"]?.stringValue
                ?: throw RynBridgeError(code = ErrorCode.INVALID_MESSAGE, message = "Missing required field: playerId")
            provider.pauseAudio(playerId)
            emptyMap()
        },
        "stopAudio" to { payload ->
            val playerId = payload["playerId"]?.stringValue
                ?: throw RynBridgeError(code = ErrorCode.INVALID_MESSAGE, message = "Missing required field: playerId")
            provider.stopAudio(playerId)
            emptyMap()
        },
        "getAudioStatus" to { payload ->
            val playerId = payload["playerId"]?.stringValue
                ?: throw RynBridgeError(code = ErrorCode.INVALID_MESSAGE, message = "Missing required field: playerId")
            val status = provider.getAudioStatus(playerId)
            status.toPayload()
        },
        "startRecording" to { payload ->
            val format = payload["format"]?.stringValue ?: "m4a"
            val quality = payload["quality"]?.stringValue ?: "medium"
            val recordingId = provider.startRecording(format, quality)
            mapOf("recordingId" to BridgeValue.string(recordingId))
        },
        "stopRecording" to { payload ->
            val recordingId = payload["recordingId"]?.stringValue
                ?: throw RynBridgeError(code = ErrorCode.INVALID_MESSAGE, message = "Missing required field: recordingId")
            val result = provider.stopRecording(recordingId)
            result.toPayload()
        },
        "pickMedia" to { payload ->
            val type = payload["type"]?.stringValue ?: "any"
            val multiple = payload["multiple"]?.boolValue ?: false
            val files = provider.pickMedia(type, multiple)
            mapOf("files" to BridgeValue.array(files.map { BridgeValue.dict(it.toPayload()) }))
        }
    )
}
